h/services/job_queue.py

```diff
@@ -0,0 +1,193 @@
+       even longer: if the job queue is currently long then the periodic task
+       may have to run multiple times before it gets to your job).
+
+       In contrast tasks added to Celery can be processed by a worker almost
+       immediately.
+
+    3. The job queue is very simple and has far fewer features than Celery.
+
+    Celery should be the default task queue for almost all tasks, and only jobs
+    that really need Postgres transactionality should use this custom job
+    queue.
+
+    At the time of writing this job queue only supports a single type of job:
+    synchronizing annotations from Postgres to Elasticsearch.
+    """
+
+    def __init__(self, db, es, batch_indexer, limit):
+        self._db = db
+        self._es = es
+        self._batch_indexer = batch_indexer
+        self._limit = limit
+
+    def add_sync_annotation_job(self, annotation_id, tag, scheduled_at=None):
+        """Add an annotation to be synced to Elasticsearch."""
+        self.add_sync_annotation_jobs([annotation_id], tag, scheduled_at)
+
+    def add_sync_annotation_jobs(self, annotation_ids, tag, scheduled_at=None):
+        """Add a list of annotations to the queue to be synced to Elasticsearch."""
+        self._db.add_all(
+            Job(
+                tag=tag,
+                scheduled_at=scheduled_at,
+                kwargs={"annotation_id": annotation_id},
+            )
+            for annotation_id in annotation_ids
+        )
+
+    def sync_annotations(self):
+        """
+        Synchronize annotations from Postgres to Elasticsearch.
+
+        This method is meant to be run periodically. It's called by a Celery
+        task that's scheduled periodically by h-periodic.
+
+        Each time this method runs it considers a fixed number of sync
+        annotation jobs from the job queue. For each job, if the annotation is
+        already present and up to date in Elasticsearch, then it removes the
+        job from the queue. If the annotation is missing from Elasticsearch or
+        out of date in Elasticsearch then the method re-syncs the annotation
+        into Elasticsearch, and leaves the job on the queue to be re-checked
+        and removed the next time the method runs.
+
+        If the DB somehow contains an annotation that always fails to index
+        into Elasticsearch, then a job for that annotation will never be
+        removed from the queue! As far as I know it isn't possible for the DB
+        to contain such an annotation.
+        """
+        jobs = self._get_jobs_from_queue()
+
+        if not jobs:
+            return
+
+        annotation_ids = [job.kwargs["annotation_id"] for job in jobs]
+        annotations_from_db = self._get_annotations_from_db(annotation_ids)
+        annotations_from_es = self._get_annotations_from_es(annotation_ids)
+
+        missing_from_db = []
+        missing_from_es = []
+        outdated_in_es = []
+        up_to_date_in_es = []
+
+        for job in jobs:
```

Comment:
I think it might be an over-optimisation. It may be easier to ignore that optimisation for the sake of keeping the code simpler. On the other hand it's probably do-able

Reply:
:+1: